zerver/lib/rate_limiter.py

```diff
@@ -569,8 +569,16 @@ def client_is_exempt_from_rate_limiting(request: HttpRequest) -> bool:
 
     # Don't rate limit requests from Django that come from our own servers,
     # and don't rate-limit dev instances
-    client = RequestNotes.get_notes(request).client
-    return (client is not None and client.name.lower() == "internal") and (
+    notes = RequestNotes.get_notes(request)
+    # Take client_name before the Client object's name to ensure edge cases like overriding client=internal in
+    # request bodies work as expected.
+    client_name = ""
+    if notes.client_name is not None:
+        client_name = notes.client_name
+    if notes.client is not None and client_name == "":
+        client_name = notes.client.name
```

Comment:
I'm not sure I understand the situation where `client` would be `None` via middleware; in `process_client`, we set setting `client` from the `client_name` unconditionally, and I would expect that to happen before this gets called. 

So I'm starting to wonder if what's actually going on here is some sort of invariant violation where `process_client` is not being called before calling rate limiting functions? Or should we be able to just assert that `notes.client_name is not None`?

Reply:
I seem to recall the calls to `get_or_create_client` (or some method to that effect) returning `None` for a client name of `internal` when stepping through in the debugger. I'll take another look Monday.